src/quark-core/src/QuarkNonceManager.sol

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.23;
+
+import {IQuarkWallet} from "quark-core/src/interfaces/IQuarkWallet.sol";
+
+/**
+ * @title Quark Nonce Manager
+ * @notice Contract for managing nonces for Quark wallets
+ * @author Compound Labs, Inc.
+ */
+contract QuarkNonceManager {
+    error NonReplayableNonce(address wallet, bytes32 nonce, bytes32 submissionToken);
+    error InvalidSubmissionToken(address wallet, bytes32 nonce, bytes32 submissionToken);
+
+    event NonceSubmitted(address wallet, bytes32 nonce, bytes32 submissionToken);
+
+    /// @notice Represents the unclaimed bytes32 value.
+    bytes32 public constant FREE = bytes32(uint256(0));
+
+    /// @notice A token that implies a Quark Operation is no longer replayable.
+    bytes32 public constant EXHAUSTED = bytes32(type(uint256).max);
+
+    /// @notice Mapping from nonces to last used submission token.
+    mapping(address wallet => mapping(bytes32 nonce => bytes32 lastToken)) public nonceSubmissions;
+
+    /**
+     * @notice Returns the nonce token (last submission token) for a given nonce. For finalized scripts, this will be `uint256(-1)`. For unclaimed nonces, this will be `uint256(0)`. Otherwise, it will be the next value in the replay chain.
+     * @param wallet The wallet for which to get the nonce token.
+     * @param nonce The nonce for the given request.
+     * @return submissionToken The last used submission token, or 0 if unused or -1 if finalized.
+     */
+    function getNonceSubmission(address wallet, bytes32 nonce) external view returns (bytes32 submissionToken) {
+        return nonceSubmissions[wallet][nonce];
+    }
+
+    /**
+     * @notice Attempts a first or subsequent submission of a given nonce from a wallet.
+     * @param nonce The nonce of the chain to submit.
+     * @param submissionToken The submission token of the submission. For single-use operations, set `submissionToken` to `uint256(-1)`. For first-use replayable operations, set `submissionToken` = `nonce`.
+     */
+    function submitNonceToken(bytes32 nonce, bytes32 submissionToken) external {
```

Comment:
Still think we should use `{verb}NonceSubmission` here to be more consistent with the rest of the naming in this contract. That verb could be `claim`, `submit`, `write`; the point is we should just pick a term (`NonceSubmission`) and use it consistently rather than have multiple ways to describe the same thing.

Reply:
`get` / `set` would be fine. Tbh I would also be ok with just `submit`: `QuarkNonceManager.submit(nonce, token)` reads fine.